arroyo-worker/src/connectors/filesystem/s3/source.rs

```diff
@@ -0,0 +1,214 @@
+            .bucket(self.bucket.clone())
+            .set_prefix(Some(self.prefix.clone()))
+            .send()
+            .await
+            .unwrap();
+        let mut prev_file_found = prev_s3_key == "";
+
+        let mut contents = list_res.contents.unwrap();
+
+        while let Some(obj) = contents.pop() {
+            let obj_key = obj.key.unwrap();
+            if obj_key.ends_with('/') {
+                continue;
+            }
+
+            // if restoring from checkpoint, skip until we find the file we were on
+            if !prev_file_found && obj_key != prev_s3_key {
+                continue;
+            }
+            prev_file_found = true;
+
+            let get_res = client
+                .get_object()
+                .bucket(self.bucket.clone())
+                .key(obj_key.clone())
+                .send()
+                .await
+                .unwrap();
+            let body = get_res.body;
+
+            if self.compression == "zstd" {
+                let r = ZstdDecoder::new(BufReader::new(body.into_async_read()));
+                match self
+                    .read_file(ctx, BufReader::new(r).lines(), &obj_key, prev_lines_read)
+                    .await
+                {
+                    Some(finish_type) => return finish_type,
+                    None => {}
+                }
+            } else if self.compression == "gzip" {
+                let r = GzipDecoder::new(BufReader::new(body.into_async_read()));
+                match self
+                    .read_file(ctx, BufReader::new(r).lines(), &obj_key, prev_lines_read)
+                    .await
+                {
+                    Some(finish_type) => return finish_type,
+                    None => {}
+                }
+            } else {
+                match self
+                    .read_file(
+                        ctx,
+                        BufReader::new(body.into_async_read()).lines(),
+                        &obj_key,
+                        prev_lines_read,
+                    )
+                    .await
+                {
+                    Some(finish_type) => return finish_type,
+                    None => {}
+                }
+            }
+        }
+        info!("S3 source finished");
+        SourceFinishType::Final
+    }
+
+    async fn read_file<R: AsyncBufRead + Unpin>(
+        &mut self,
+        ctx: &mut Context<(), T>,
+        mut reader: Lines<R>,
+        obj_key: &str,
+        prev_lines_read: usize,
+    ) -> Option<SourceFinishType> {
+        let mut lines_read = 0;
+        while let Ok(Some(s)) = reader.next_line().await {
+            // if we're restoring from checkpoint, skip until we find the line we were on
+            if lines_read < prev_lines_read {
+                lines_read += 1;
+                continue;
+            }
+            // json! will correctly handle escaping
+            let value = serde_json::from_value(serde_json::json!({
```

Comment:
Rather than manually mapping it to raw_string, you should be able to use the Format attached to the table. See the Kafka source for an example.